File: proga/1_semestr/labs/lab3/src/main/java/dev/lysmux/lab3/entity/ability/EmitRays.java
package dev.lysmux.lab3.entity.ability;

import dev.lysmux.lab3.entity.ray.Ray;

import java.util.List;
import java.util.Objects;

public class EmitRays extends Ability {
    protected Ray emmitedRay;
    protected List<Ray> influencingRays;

    public EmitRays(Ray emmitedRay, Ray... influencingRays) {;
        this.emmitedRay = emmitedRay;
        this.influencingRays = List.of(influencingRays);
    }

    public String influenced() {
        List<String> ray_kinds = influencingRays.stream().map(ray -> ray.getKind().getValue()).toList();

        return "под влиянием " + switch (ray_kinds.size()) {
            case 1 -> influencingRays.get(0).toString();
            case 2 -> String.join("или", ray_kinds.subList(0, 1));
            default -> String.join(", ", ray_kinds.subList(0, ray_kinds.size() - 1))
                    + " или " + ray_kinds.get(ray_kinds.size() - 1);
        } + " лучей";
    }

    @Override
    public String getDescription() {
        return String.join(
                " ",
                "испускать",
                emmitedRay.getVisibility().getValue(),
                emmitedRay.getPluralName(),
                emmitedRay.getKind().getValue(),
                "даже",
                influenced()
        );
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        EmitRays emitRays = (EmitRays) o;
        return Objects.equals(emmitedRay, emitRays.emmitedRay)
                && Objects.equals(influencingRays, emitRays.influencingRays);
    }

    @Override
    public int hashCode() {
        return Objects.hash(emmitedRay, influencingRays);
    }

    @Override
    public String toString() {
        return "испускать лучи";
    }
}

File: proga/1_semestr/labs/lab3/src/main/kotlin/dev/lysmux/lab3/entity/ability/EmitRays.kt
package dev.lysmux.lab3.entity.ability

import dev.lysmux.lab3.entity.ray.Ray

class EmitRays(private val emmitedRay: Ray, private vararg val influencingRays: Ray) : Ability() {
    private fun influenced(): String {
        val rayKindNames = influencingRays.map { it.kind.value }
        val builder = StringBuilder("под влиянием ")

        when (rayKindNames.size) {
            1 -> builder.append(rayKindNames[0])
            2 -> builder.append("${rayKindNames[0]} или ${rayKindNames[1]}")
            else -> {
                builder.append(rayKindNames.subList(0, rayKindNames.size - 1).joinToString(", "))
                builder.append(" или ${rayKindNames.last()}")
            }
        }
        builder.append(" лучей")
        return builder.toString()
    }

    override fun getDescription(): String {
        return "испускать ${emmitedRay.visibility.value} ${emmitedRay.getPluralName()} ${
            emmitedRay.kind.value
        } даже ${influenced()}"
    }

    override fun toString() = "испускать лучи"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is EmitRays) return false

        if (emmitedRay != other.emmitedRay) return false
        if (!influencingRays.contentEquals(other.influencingRays)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = emmitedRay.hashCode()
        result = 31 * result + influencingRays.contentHashCode()
        return result
    }
}